### PR TITLE
ISSUE 20: Suppression List SDK

### PR DESCRIPTION
**Pull request by aydrian**
Please review and discuss.

### CODE REVIEW

File: test/spec/suppressionList.spec.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+  before(function () {
+    // setting up a client for all tests to use
+    var key = '12345678901234567890';
+
+    client = new SparkPost(key);
+    suppressionList = require('../../lib/suppressionList')(client);
+  });
+
+  it('should expose a public search method', function () {
```
Comment (colestrode) [inline test/spec/suppressionList.spec.js]: I'm not a fan of `.to.be.a.function` style tests. I feel like testing the functionality of these methods covers this

File: test/spec/suppressionList.spec.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+chai.use(sinonChai);
+
+describe('Suppression List Library', function() {
+  var client, suppressionList;
+
+  before(function () {
+    // setting up a client for all tests to use
+    var key = '12345678901234567890';
+
+    client = new SparkPost(key);
```
Comment (colestrode) [inline test/spec/suppressionList.spec.js]: I would just mock the client since this is a unit test. Then in the below tests all you will need to test is that you are calling the client correctly. Making requests and formatting the response is covered in the client's unit tests.

File: test/spec/suppressionList.spec.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      nock('https://api.sparkpost.com')
+        .get('/api/v1/suppression-list?limit=5')
+        .reply(200, { ok: true });
+
+      suppressionList.search({limit: 5}, function(err, data) {
+        // need to make sure we called get method
+        expect(requestSpy.calledOnce).to.be.true;
+
+        // making sure the correct uri was constructed
+        expect(data.res.request.uri.href).to.equal('https://api.sparkpost.com:443/api/v1/suppression-list?limit=5');
```
Comment (colestrode) [inline test/spec/suppressionList.spec.js]: url construction is the purview of the client, not the suppressionList, so no need to test that here

Comment (aydrian) [inline test/spec/suppressionList.spec.js]: But for the search Method, what you set as params gets added to the querystring.

File: test/spec/suppressionList.spec.js (hunk 1 above)
Comment (aydrian) [inline test/spec/suppressionList.spec.js]: I got this from whoever wrote the tests for the transmissions library. :)

File: test/spec/suppressionList.spec.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+  it('should expose a public removeStatus method', function () {
+    expect(suppressionList.removeStatus).to.be.a.function;
+  });
+
+  it('should expose a public upsert method', function () {
+    expect(suppressionList.upsert).to.be.a.function;
+  });
+
+  describe('search Method', function() {
```
Comment (colestrode) [inline test/spec/suppressionList.spec.js]: I would add some unhappy path tests below: bad/missing parameters, empty strings, etc

File: lib/suppressionList.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  var model = {
+    recipients: input
+  };
+
+  return model;
+};
+
+module.exports = function(client) {
+
+  return {
```
Comment (colestrode) [inline lib/suppressionList.js]: I'd think you'd want more error checking

**Comment by bdeanindy** [comment]
All of Cole's suggestions are valid.

Testing behavior over structure is a good way to think about the tests. I especially agree with authoring tests for the "unhappy" paths. We know what we expect to be good, but want to make sure future PRs do not break things we don't want broken (integration tests).

**Comment by aydrian** [comment]
Good idea. Do you have a good example of that? In my experience, you don't want to be too strict.
